File: domain-implementation/src/main/java/com/mrezanasirloo/domain.implementation/repository/RepositoryID3Tag.java
package com.mrezanasirloo.domain.implementation.repository;

import android.support.annotation.NonNull;

import com.mrezanasirloo.domain.model.SongDomain;
import com.mrezanasirloo.domain.model.SongTagDomain;

import org.jaudiotagger.audio.AudioFile;
import org.jaudiotagger.audio.AudioFileIO;
import org.jaudiotagger.audio.exceptions.CannotReadException;
import org.jaudiotagger.audio.exceptions.InvalidAudioFrameException;
import org.jaudiotagger.audio.exceptions.ReadOnlyFileException;
import org.jaudiotagger.tag.FieldKey;
import org.jaudiotagger.tag.Tag;
import org.jaudiotagger.tag.TagException;

import java.io.File;
import java.io.IOException;

import javax.inject.Inject;
import javax.inject.Singleton;

import io.reactivex.Observable;

@Singleton
public class RepositoryID3Tag {
    private static final String TAG = RepositoryID3Tag.class.getSimpleName();

    @Inject
    public RepositoryID3Tag() {
    }

    @NonNull
    private AudioFile getAudioFile(@NonNull String path) throws TagException, ReadOnlyFileException, CannotReadException, InvalidAudioFrameException, IOException {
        return AudioFileIO.read(new File(path));
    }

    private SongTagDomain getSongId3Tag(@NonNull SongDomain songDomain) throws ReadOnlyFileException, IOException, TagException, InvalidAudioFrameException, CannotReadException {
        Tag tag = getAudioFile(songDomain.getData()).getTagOrCreateAndSetDefault();
        return new SongTagDomain(
                new SongDomain(
                        songDomain.getId(),
                        tag.getFirst(FieldKey.TITLE),
                        Integer.valueOf(tag.getFirst(FieldKey.TRACK)),
                        Integer.valueOf(tag.getFirst(FieldKey.YEAR)),
                        songDomain.getDuration(),
                        songDomain.getData(),
                        songDomain.getDateModified(),
                        songDomain.getAlbumId(),
                        tag.getFirst(FieldKey.ALBUM),
                        songDomain.getArtistId(),
                        tag.getFirst(FieldKey.ARTIST)
                ),
                tag.getFirst(FieldKey.GENRE),
                tag.getFirstArtwork().getBinaryData()
        );
    }

    public Observable<SongTagDomain> readId3Tag(SongDomain song) {
        return Observable.create(emitter -> emitter.onNext(getSongId3Tag(song)));
    }
}

File: domain-implementation/src/main/java/com/mrezanasirloo/domain.implementation/repository/RepositoryID3Tag.kt
package com.mrezanasirloo.domain.implementation.repository

import android.content.Context
import android.media.MediaScannerConnection
import com.mrezanasirloo.domain.model.SongDomain
import com.mrezanasirloo.domain.model.SongTagDomain
import io.reactivex.Observable
import org.jaudiotagger.audio.AudioFileIO
import org.jaudiotagger.tag.FieldKey
import java.io.File
import javax.inject.Inject

class RepositoryID3Tag @Inject constructor(
        val context: Context
) {
    @Throws(Exception::class)
    private fun getSongId3Tag(songDomain: SongDomain): SongTagDomain {
        val tag = AudioFileIO.read(File(songDomain.data)).tagOrCreateAndSetDefault
        return SongTagDomain(
                SongDomain(
                        -1,
                        songDomain.id,
                        tag.getFirst(FieldKey.TITLE),
                        songDomain.trackNumber,
                        songDomain.year,
                        songDomain.duration,
                        songDomain.data,
                        songDomain.dateModified,
                        songDomain.albumId,
                        tag.getFirst(FieldKey.ALBUM),
                        songDomain.artistId,
                        tag.getFirst(FieldKey.ARTIST)
                ),
                tag.getFirst(FieldKey.GENRE),
                tag.getFirst(FieldKey.YEAR),
                tag.getFirst(FieldKey.TRACK),
                tag.firstArtwork.binaryData
        )
    }

    @Throws(Exception::class)
    private fun updateSongId3Tag(song: SongTagDomain): SongTagDomain {
        val songDomain = song.songDomain
        val audioFile = AudioFileIO.read(File(songDomain.data))
        val tag = audioFile.tagOrCreateAndSetDefault
        tag.setField(FieldKey.TITLE, songDomain.title)
        tag.setField(FieldKey.ALBUM, songDomain.albumName)
        tag.setField(FieldKey.ARTIST, songDomain.artistName)
        tag.setField(FieldKey.TRACK, song.track)
        tag.setField(FieldKey.GENRE, song.genre)
        tag.setField(FieldKey.YEAR, song.year)
        try {
            audioFile.commit()
        } catch (e: Exception) {
            e.printStackTrace()
        }
        MediaScannerConnection.scanFile(context.applicationContext, listOf(song.songDomain.data).toTypedArray(), null, null)
        return song
    }

    fun updateId3Tag(song: SongTagDomain): Observable<SongTagDomain> {
        return Observable.create { emitter -> emitter.onNext(updateSongId3Tag(song)) }
    }

    fun readId3Tag(song: SongDomain): Observable<SongTagDomain> {
        return Observable.create { emitter -> emitter.onNext(getSongId3Tag(song)) }
    }

    companion object {
        private val TAG = RepositoryID3Tag::class.java.simpleName
    }
}
